string_manipulation.rb

```diff
@@ -1,30 +1,133 @@
+# Helpers
+def swap_in_place(string, left, right) #time complexity = O(c), space = O(c) (temp_var)
+  temp_var = string[left]
+  string[left] = string[right]
+  string[right] = temp_var
+end
+
+
 # A method to reverse a string in place.
 def string_reverse(my_string)
-  puts "NOT IMPLEMENTED"
+  return my_string if my_string.length < 2
+
+  i = 0
+
+  while i < my_string.length/2
+    swap_in_place(my_string, i, my_string.length-i-1)
+    i +=1
+  end
+
+  return my_string
 end
 
 # A method to reverse each word in a sentence, in place.
-def reverse_words(my_words)
-  puts "NOT IMPLEMENTED"
+def reverse_words(my_words) #with string_reverse
+  return my_words if my_words.length < 2
+  return string_reverse(my_words) if !my_words.include?(" ")
+
+  first = 0
+
+  my_words.length.times do |last |
+    if my_words[last].match(/\s/)
+      my_words[first...last] = string_reverse(my_words[first...last])
+      first = last + 1
+    elsif last == (my_words.length - 1)
+      my_words[first..last] = string_reverse(my_words[first..last])
+    end
+  end
+
+  return my_words
+end
+
+def reverse_words2(my_words) #without #string_reverse
+  return my_words if my_words.length < 2
+  return string_reverse(my_words) if !my_words.include?(" ")
+
+  first = 0
+
+  my_words.length.times do |last |
+    if my_words[last].match(/\s/)
+      i = first
+      j = last - 1
+      while i < j
+        swap_in_place(my_words, i, j)
+        i +=1
+        j -=1
+      end
+      first = last + 1
+    elsif last == (my_words.length - 1)
+      while first < last
+        swap_in_place(my_words, first, last)
+        first +=1
+        last -=1
+      end
+    end
+  end
+
+  return my_words
 end
 
 # A method to reverse the words in a sentence, in place.
 def reverse_sentence(my_sentence)
-  puts "NOT IMPLEMENTED"
+  my_sentence = reverse_words(my_sentence)

```

Comment:
This will also be a O(n) algorithm with O(1) space complexity. Similar notes to reverse_words.